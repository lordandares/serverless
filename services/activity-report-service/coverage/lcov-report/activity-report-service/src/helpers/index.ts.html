<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for activity-report-service/src/helpers/index.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">activity-report-service/src/helpers</a> index.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>34/34</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>18/18</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20</td><td class="line-coverage quiet"><span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import * as s3 from './s3'
&nbsp;
export { buildLocationSummaryPdf } from './build-location-summary-pdf'
export { buildShiftSummaryPdf } from './build-shift-summary-pdf'
export { buildEmail } from './build-email'
export { buildShiftEmail } from './build-shift-email'
export { getShiftDetails } from './get-shift-details'
export { getShiftReportLinks } from './get-shift-report-links'
export { getUserDetails } from './get-user-details'
export { concatPdfs } from './concat-pdfs'
export { downloadFiles } from './download-files'
export { fetchApplications } from './fetch-applications'
export { fetchLocations } from './fetch-locations'
export { getLocationReportLinks } from './get-location-report-links'
export { getShiftSummaryData } from './get-shift-summary-data'
export { getSummaryData } from './get-summary-data'
export { getFormS3Paths } from './get-references'
export { s3 }
export { sendEmail } from './send-email'
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Jan 13 2020 21:52:44 GMT+0000 (Coordinated Universal Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
