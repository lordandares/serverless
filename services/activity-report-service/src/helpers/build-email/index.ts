import * as BPromise from 'bluebird'
import {
  filter,
  find,
  flow,
  get,
  getOr,
  isEmpty,
  reject,
  toUpper,
  trim,
} from 'lodash/fp'
import * as moment from 'moment-timezone'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

import { Location } from '../../handlers/email-processor'
import { getLocationReportLinks } from '../get-location-report-links'
import { UserDetails } from '../get-user-details'

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(true, 'options.formSubmissions'),
)

const getNoActivityOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(false, 'options.skipNoActivity'),
)

export interface BuildEmail extends UserDetails {
  locations: Location[]
}

export async function buildEmail(options: BuildEmail) {
  try {
    const {
      applicationName,
      email,
      firstName,
      locations,
      preferences,
      timezone,
    } = options

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-location-reports'
      : 'daily-location-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment()
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)
    const skipNoActivity = getNoActivityOption(preferences)

    const links = await BPromise.map(locations, location =>
      getLocationReportLinks({
        location,
        withFormSubmissions,
      }),
    )

    const activeLocations = filter('link')(links)
    const inactiveLocations = reject('link')(links)

    const hasNoLocations =
      isEmpty(activeLocations) && isEmpty(inactiveLocations)

    if (
      isEmpty(activeLocations) &&
      !isEmpty(inactiveLocations) && skipNoActivity
    ) {
      console.info('skipping activity report', {
        activeLocations: activeLocations.length,
        inactiveLocations: inactiveLocations.length,
        skipNoActivity,
      })
      return false
    }

    const name = trim(firstName)

    const data = JSON.stringify({
      activeLocations,
      applicationName,
      date: dateText,
      hasNoLocations,
      inactiveLocations,
      isProduction,
      name,
      sendInactive: !skipNoActivity,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, options })
    return false
  }
}
