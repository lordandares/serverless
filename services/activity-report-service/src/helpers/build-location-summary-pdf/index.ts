import { buildActivityPdf } from '@lighthouse/common'
import * as dateFns from 'date-fns'

import { buildConfig, createSummaryPdf } from '../create-summary-pdf'

export async function buildLocationSummaryPdf(data) {
  const config = buildConfig(data)
  const { data: summaryData, pdfOptions } = config

  const pdfDefinition = await buildActivityPdf(pdfOptions, summaryData)
  const s3FullPath = getLocationSummaryPdfPath(data)
  await createSummaryPdf(pdfDefinition, s3FullPath)

  return s3FullPath
}

function getLocationSummaryPdfPath({ location }) {
  const date = new Date()
  const year = dateFns.format(date, 'YYYY')
  const month = dateFns.format(date, 'MM')
  const day = dateFns.format(date, 'DD')
  const unix = date.getTime()

  const { _id: locationId } = location
  const s3PathName = `activity-reports/${locationId}/${year}/${month}`
  const s3FileName = `${year}-${month}-${day}-summary-${unix}.pdf`
  return `${s3PathName}/${s3FileName}`
}
