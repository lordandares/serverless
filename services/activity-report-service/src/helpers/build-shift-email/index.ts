import * as BPromise from 'bluebird'
import {
  filter,
  isEmpty,
  reject,
  sortBy,
  toUpper,
  trim,
  getOr,
  flow,
  find,
  get,
} from 'lodash/fp'
import { orderBy } from 'lodash'
import * as moment from 'moment-timezone'

import { getShiftReportLinks } from '../get-shift-report-links'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-shift-report' }),
  getOr(true, 'options.formSubmissions'),
)

export async function buildShiftEmail({ userDetails, datetime }) {
  try {
    const {
      applicationName,
      email,
      firstName,
      preferences,
      timezone,
      shifts,
    } = userDetails

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-shift-reports'
      : 'daily-shift-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment(datetime)
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)

    const shiftLinks = await BPromise.map(shifts, shift =>
      getShiftReportLinks({
        datetime,
        shift,
        withFormSubmissions,
      }),
    )

    const activeShifts = orderBy(
      filter('link')(shiftLinks),
      ['lastName', 'timestamp'],
      ['asc'],
    )

    const hasNoShifts = isEmpty(activeShifts)
    const name = trim(firstName)

    const data = JSON.stringify({
      activeShifts,
      applicationName,
      date: dateText,
      hasNoShifts,
      isProduction,
      name,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, userDetails })
    return false
  }
}
