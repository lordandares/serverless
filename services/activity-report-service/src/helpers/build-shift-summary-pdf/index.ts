import { buildShiftPdf } from '@lighthouse/common'
import * as dateFns from 'date-fns'

import { buildConfig, createSummaryPdf } from '../create-summary-pdf'

export async function buildShiftSummaryPdf(data) {
  const config = buildConfig(data)
  const { data: summaryData, pdfOptions } = config

  const pdfDefinition = await buildShiftPdf(pdfOptions, summaryData)
  const s3FullPath = getShiftSummaryPdfPath(data)
  await createSummaryPdf(pdfDefinition, s3FullPath)

  return s3FullPath
}

function getShiftSummaryPdfPath({ applicationId, shift }) {
  const date = new Date()
  const year = dateFns.format(date, 'YYYY')
  const month = dateFns.format(date, 'MM')
  const day = dateFns.format(date, 'DD')
  const unix = date.getTime()

  const { _id: shiftId } = shift
  const s3PathName = `activity-reports/shifts/${applicationId}/${year}/${month}/${day}`
  const s3FileName = `${year}-${month}-${day}-shift-${shiftId}-summary-${unix}.pdf`
  return `${s3PathName}/${s3FileName}`
}
