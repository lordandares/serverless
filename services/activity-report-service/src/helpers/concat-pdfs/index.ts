import { each } from 'lodash/fp'
import * as uuid from 'node-uuid'

import * as hummus from '@lighthouse/hummus-lambda'

export async function concatPdfs(filePaths, directoryPath) {
  const fileName = 'output.pdf'
  const filePath = `${directoryPath}/${fileName}`

  const pdfWriter = hummus.createWriter(filePath)
  each(path => appendPDFPageFromPDFWithAnnotations(pdfWriter, path), filePaths)
  pdfWriter.end()

  return filePath
}

/* istanbul ignore next */
function appendPDFPageFromPDFWithAnnotations(pdfWriter, sourcePDFPath) {
  const cpyCxt = pdfWriter.createPDFCopyingContext(sourcePDFPath)
  const cpyCxtParser = cpyCxt.getSourceDocumentParser()

  for (let i = 0; i < cpyCxtParser.getPagesCount(); ++i) {
    const pageDictionary = cpyCxtParser.parsePageDictionary(i)

    if (!pageDictionary.exists('Annots')) {
      // no annotation. append as is
      cpyCxt.appendPDFPageFromPDF(i)
    } else {
      // this let here will save any reffed objects from the copied annotations object.
      // they will be written after the page copy writing as to not to disturb the
      // page object writing itself.
      let reffedObjects

      pdfWriter.getEvents().once('OnPageWrite', params => {
        // using the page write event, write the new annotations. just copy the object
        // as is, saving any referenced objects for future writes
        params.pageDictionaryContext.writeKey('Annots')
        reffedObjects = cpyCxt.copyDirectObjectWithDeepCopy(
          pageDictionary.queryObject('Annots'),
        )
      })
      // write page. this will trigger the event
      cpyCxt.appendPDFPageFromPDF(i)
      // now write the reffed object (should be populated cause onPageWrite was written)
      // note that some or all annotations may be embedded, in which case this array
      // wont hold all annotation objects
      if (reffedObjects && reffedObjects.length > 0) {
        cpyCxt.copyNewObjectsForDirectObject(reffedObjects)
      }
    }
  }
}
