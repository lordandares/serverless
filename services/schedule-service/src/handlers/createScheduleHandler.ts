import { errors, http, schemas } from '@lighthouse/serverless-common'
import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'
import { ScheduleService } from '../service/ScheduleService'

export async function createScheduleHandler(
  event: APIGatewayProxyEvent,
): Promise<APIGatewayProxyResult> {
  try {
    const { body: eventBody, headers } = event

    if (!eventBody) {
      throw new errors.ApplicationError({
        message: 'CreateScheduleHandler: missing body on event',
      })
    }

    if (!headers) {
      throw new errors.ApplicationError({
        message: 'CreateScheduleHandler: missing headers on event',
      })
    }

    const body: schemas.SchedulePayloadSchema = JSON.parse(eventBody)
    const applicationId: string = headers['lio-application-id']
    const userId: string = headers['lio-user-id']

    const schedule: schemas.SchedulePayloadSchema = await ScheduleService.createSchedule(
      {
        applicationId,
        body,
        userId,
      },
    )

    return {
      body: JSON.stringify(schedule),
      statusCode: http.STATUS_CODES.SUCCESSFUL.CREATED,
    }
  } catch (err) {
    console.error('CreateScheduleError', {
      err,
      event,
    })

    return errors.httpErrorHandler(err)
  }
}
