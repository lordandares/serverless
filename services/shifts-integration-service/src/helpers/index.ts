export { default as getPlatformEndpoint } from './get-platform-endpoint'
export { default as getSnsPayload } from './get-sns-payload'
export {
  getWinTeamMessages,
  getWinTeamTimePunch,
} from './get-winteam-time-punch'
export { default as notifyUser } from './notify-user'
export { default as parseWinTeamMessages } from './parse-winteam-messages'
export { default as sendMessage } from './send-message'
export {
  default as validateEndpointAttributes,
} from './validate-endpoint-attributes'
