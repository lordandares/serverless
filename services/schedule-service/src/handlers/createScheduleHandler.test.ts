import { mongo, schemas } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk-mock'
import * as MockDate from 'mockdate'

import { createApiEvent } from '../../../../__test__/helpers'
import {
  areaDocument,
  scheduleDocument,
  schedulePayload,
} from '../__fixtures__'
import { createScheduleHandler } from './createScheduleHandler'

jest.mock('../service/lib/transform', () => {
  const {
    scheduleDocument: document,
    schedulePayload: payload,
  } = require('../__fixtures__')

  return {
    documentToSchedulePayload: () => payload,
    payloadToScheduleDocument: () => document,
  }
})

beforeEach(() => {
  process.env.CREATE_SCHEDULE_STEP_FUNCTION = 'create-schedule-step-function'
  process.env.IS_OFFLINE = 'true'
  process.env.TABLE_SCHEDULES = 'table-schedules'

  MockDate.set('2019-10-14T00:00:00.000Z')
})

afterEach(() => MockDate.reset())

test('event body error', async () => {
  expect.assertions(1)

  const result = await createScheduleHandler(
    createApiEvent({
      body: null,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ApplicationError\\",\\"status\\":500,\\"message\\":\\"CreateScheduleHandler: missing body on event\\"}",
  "statusCode": 500,
}
`)
})

test('event headers error', async () => {
  expect.assertions(1)

  const result = await createScheduleHandler(
    createApiEvent({
      body: {},
      headers: null,
    }),
  )

  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ApplicationError\\",\\"status\\":500,\\"message\\":\\"CreateScheduleHandler: missing headers on event\\"}",
  "statusCode": 500,
}
`)
})

test('event applicationId header error', async () => {
  expect.assertions(1)

  const result = await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': '',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ApplicationError\\",\\"status\\":500,\\"message\\":\\"The \`applicationId\` is missing from headers\\"}",
  "statusCode": 500,
}
`)
})

test('event userId header error', async () => {
  expect.assertions(1)

  const result = await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': '',
      },
    }),
  )

  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ApplicationError\\",\\"status\\":500,\\"message\\":\\"The \`userId\` is missing from headers\\"}",
  "statusCode": 500,
}
`)
})

test('event body error', async () => {
  expect.assertions(1)

  const result = await createScheduleHandler(
    createApiEvent({
      body: {},
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ValidationError\\",\\"status\\":400,\\"message\\":\\"A schedule body is required\\"}",
  "statusCode": 400,
}
`)
})

test('generic error', async () => {
  expect.assertions(2)

  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockRejectedValue(new Error('NetworkError'))

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)

  const result = await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(putSpy).toHaveBeenCalled()
  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ApplicationError\\",\\"status\\":500,\\"message\\":\\"NetworkError\\"}",
  "statusCode": 500,
}
`)

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('application error', async () => {
  expect.assertions(1)

  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const validateSpy = jest.spyOn(schemas, 'validate')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  validateSpy.mockImplementation(
    jest
      .fn()
      .mockResolvedValueOnce({})
      .mockRejectedValueOnce(new Error('Schedule has malformed data')),
  )

  const result = await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(result).toMatchInlineSnapshot(`
Object {
  "body": "{\\"name\\":\\"ApplicationError\\",\\"status\\":500,\\"message\\":\\"Schedule has malformed data\\"}",
  "statusCode": 500,
}
`)

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()
  validateSpy.mockRestore()
})

test('validates against payload schema', async () => {
  expect.assertions(1)

  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({ TableName: 'schedule-table' })
  const startExecutionSpy = jest.fn().mockResolvedValue({})
  const validateSpy = jest.spyOn(schemas, 'validate')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('StepFunctions', 'startExecution', startExecutionSpy)

  await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(validateSpy).toHaveBeenCalledWith({
    data: schedulePayload,
    schema: schemas.schedulePayloadSchema,
  })

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()
  validateSpy.mockRestore()

  AWS.restore()
})

test('validates against document schema', async () => {
  expect.assertions(1)

  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const startExecutionSpy = jest.fn().mockResolvedValue({})
  const validateSpy = jest.spyOn(schemas, 'validate')

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('StepFunctions', 'startExecution', startExecutionSpy)

  await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(validateSpy).toHaveBeenCalledWith({
    data: scheduleDocument,
    options: { context: { isNew: true } },
    schema: schemas.scheduleDocumentSchema,
  })

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()
  validateSpy.mockRestore()

  AWS.restore()
})

test('returns 201 with new document', async () => {
  expect.assertions(4)

  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const startExecutionSpy = jest.fn().mockResolvedValue({})

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(areaDocument),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('StepFunctions', 'startExecution', startExecutionSpy)

  const result = await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(startExecutionSpy.mock.calls).toMatchSnapshot()
  expect(putSpy.mock.calls).toMatchSnapshot()
  expect(startExecutionSpy.mock.calls).toMatchSnapshot()
  expect(result).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})

test('returns 201 with new document when locations missing', async () => {
  expect.assertions(1)

  const getCollectionSpy = jest.spyOn(mongo, 'getCollection')
  const objectIdSpy = jest.spyOn(mongo, 'ObjectId')
  const putSpy = jest.fn().mockResolvedValue({})
  const startExecutionSpy = jest.fn().mockResolvedValue({})

  getCollectionSpy.mockImplementation(
    jest.fn().mockResolvedValue({
      findOne: jest.fn().mockResolvedValue(undefined),
    }),
  )
  objectIdSpy.mockImplementation(jest.fn().mockReturnValue('id'))

  AWS.mock('DynamoDB.DocumentClient', 'put', putSpy)
  AWS.mock('StepFunctions', 'startExecution', startExecutionSpy)

  const result = await createScheduleHandler(
    createApiEvent({
      body: schedulePayload,
      headers: {
        'lio-application-id': 'application1',
        'lio-user-id': 'user1',
      },
    }),
  )

  expect(result).toMatchSnapshot()

  getCollectionSpy.mockRestore()
  objectIdSpy.mockRestore()

  AWS.restore()
})
