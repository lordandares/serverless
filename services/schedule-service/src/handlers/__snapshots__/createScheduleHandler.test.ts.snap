// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns 201 with new document 1`] = `
Array [
  Array [
    Object {
      "input": "{\\"applicationId\\":\\"application1\\",\\"locations\\":[\\"location1\\"],\\"scheduleId\\":\\"schedule1\\"}",
      "name": "application1-schedule1-1571011200000",
      "stateMachineArn": "create-schedule-step-function",
    },
    [Function],
  ],
]
`;

exports[`returns 201 with new document 2`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "applicationId": "application1",
        "createdAt": "2019-10-13T00:00:00.000Z",
        "createdBy": Object {
          "id": "user1",
          "label": "Unknown User",
          "type": "user",
        },
        "data": Object {
          "areas": Array [
            "area1",
          ],
          "enabled": true,
          "included": Object {},
          "locations": Array [
            "location1",
          ],
          "name": "Schedule One",
          "serviceHours": Object {
            "hours": Array [
              Object {
                "description": "MON 11:00 - TUE 00:00",
                "end": 86400000,
                "start": 39600000,
                "type": "DEFAULT",
              },
            ],
            "timezone": "Australia/Melbourne",
          },
          "strategy": Object {
            "options": Object {
              "duration": Object {
                "unit": "minute",
                "value": 15,
              },
              "frequency": Object {
                "unit": "minute",
                "value": 15,
              },
            },
            "type": "stopwatch",
          },
          "type": "visit",
        },
        "endAt": "2019-10-1500:00:00.000Z",
        "groupType": "schedule",
        "pk": "application1-schedule",
        "scheduleId": "schedule1",
        "sk": "2019-10-13T00:00:00.000Z-schedule1",
        "startAt": "2019-10-1400:00:00.000Z",
        "updatedAt": "2019-10-13T01:00:00.000Z",
        "updatedBy": Object {
          "id": "user1",
          "label": "Unknown User",
          "type": "user",
        },
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;

exports[`returns 201 with new document 3`] = `
Array [
  Array [
    Object {
      "input": "{\\"applicationId\\":\\"application1\\",\\"locations\\":[\\"location1\\"],\\"scheduleId\\":\\"schedule1\\"}",
      "name": "application1-schedule1-1571011200000",
      "stateMachineArn": "create-schedule-step-function",
    },
    [Function],
  ],
]
`;

exports[`returns 201 with new document 4`] = `
Object {
  "body": "{\\"areas\\":[\\"area1\\"],\\"enabled\\":true,\\"endAt\\":\\"2019-10-1500:00:00.000Z\\",\\"locations\\":[\\"location1\\"],\\"name\\":\\"Schedule One\\",\\"serviceHours\\":{\\"hours\\":[{\\"description\\":\\"MON 11:00 - TUE 00:00\\",\\"end\\":86400000,\\"start\\":39600000,\\"type\\":\\"DEFAULT\\"}],\\"timezone\\":\\"Australia/Melbourne\\"},\\"startAt\\":\\"2019-10-1400:00:00.000Z\\",\\"strategy\\":{\\"options\\":{\\"duration\\":{\\"unit\\":\\"minute\\",\\"value\\":15},\\"frequency\\":{\\"unit\\":\\"minute\\",\\"value\\":15}},\\"type\\":\\"stopwatch\\"},\\"type\\":\\"visit\\"}",
  "statusCode": 201,
}
`;

exports[`returns 201 with new document when locations missing 1`] = `
Object {
  "body": "{\\"areas\\":[\\"area1\\"],\\"enabled\\":true,\\"endAt\\":\\"2019-10-1500:00:00.000Z\\",\\"locations\\":[\\"location1\\"],\\"name\\":\\"Schedule One\\",\\"serviceHours\\":{\\"hours\\":[{\\"description\\":\\"MON 11:00 - TUE 00:00\\",\\"end\\":86400000,\\"start\\":39600000,\\"type\\":\\"DEFAULT\\"}],\\"timezone\\":\\"Australia/Melbourne\\"},\\"startAt\\":\\"2019-10-1400:00:00.000Z\\",\\"strategy\\":{\\"options\\":{\\"duration\\":{\\"unit\\":\\"minute\\",\\"value\\":15},\\"frequency\\":{\\"unit\\":\\"minute\\",\\"value\\":15}},\\"type\\":\\"stopwatch\\"},\\"type\\":\\"visit\\"}",
  "statusCode": 201,
}
`;
